TypeError: Cannot prevent extension of an object with external array elements

#### Issue Description
Object.freeze on some arrays causes exception to be thrown:

```
TypeError: Cannot prevent extension of an object with external array elements
```

Looks like possibly a v8 bug: https://bugs.chromium.org/p/v8/issues/detail?id=4460

#### Environment
- node version: 4.4.4
- trails version: master
- operating system: all?

Another error (crash) I just discovered, I'm working on an example project with Trails and Angular 2 but the freeze capabilities prevent my code to work because angular2 can't be `require` so I have this error : 
```
error: 
 Error: Cannot find module 'angular2'
    at Function.Module._resolveFilename (module.js:325:15)
    at Function.Module._load (module.js:276:25)
    at Module.require (module.js:353:17)
    at require (internal/module.js:12:17)
    at /Users/jaumard/IdeaProjects/trails-example-express-gulp-angular2/node_modules/trails/lib/trails.js:119:53
    at Array.find (native)
    at /Users/jaumard/IdeaProjects/trails-example-express-gulp-angular2/node_modules/trails/lib/trails.js:119:36
    at Array.forEach (native)
    at Object.freezeConfig (/Users/jaumard/IdeaProjects/trails-example-express-gulp-angular2/node_modules/trails/lib/trails.js:117:15)
    at TrailsApp.<anonymous> (/Users/jaumard/IdeaProjects/trails-example-express-gulp-angular2/node_modules/trails/lib/trails.js:169:59)
    at TrailsApp.g (events.js:260:16)
    at emitNone (events.js:72:20)
    at TrailsApp.emit (events.js:166:7)
    at TrailsApp.emit (/Users/jaumard/IdeaProjects/trails-example-express-gulp-angular2/node_modules/trails/index.js:142:23)
    at /Users/jaumard/IdeaProjects/trails-example-express-gulp-angular2/node_modules/trails/lib/trailpack.js:34:23
    at process._tickDomainCallback (node.js:407:9)
    at Function.Module.runMain (module.js:443:11)
    at startup (node.js:134:18)
    at node.js:962:3
error: The application did not boot successfully.
```

The error come from here https://github.com/trailsjs/trails/blob/master/lib/trails.js#L119 when it try to do `require('angular2')`

No angular2 is not 'requirable' at all, in any js files you can't do `require('angular2')` or you have ` Cannot find module 'angular2'`. It's another issue than this one but I put it here because it's related.

--- a/index.js
+++ b/index.js
@@ -66,7 +66,7 @@ module.exports = class TrailsApp extends events.EventEmitter {
       },
       loadedModules: {
         enumerable: false,
-        value: lib.Trails.getExternalModules(this.pkg)
+        value: lib.Trails.getExternalModules(this.pkg, this.config.main.paths.root)
       }
     })
 

--- a/lib/trails.js
+++ b/lib/trails.js
@@ -60,9 +60,6 @@ module.exports = {
 
     const nestedEnvs = lib.Trails.getNestedEnv(config)
     if (nestedEnvs.length) {
-      nestedEnvs.forEach(key => {
-        console.error('env config [', key ,'] contains the "env" property')
-      })
       throw new lib.Errors.ConfigValueError('Environment configs cannot contain an "env" property')
     }
 
@@ -77,17 +74,13 @@ module.exports = {
    * the config object will contain trailpack modules, and we do not necessarily
    * want to deep freeze all of them. In general, messing directly with loaded
    * modules is not safe.
-   *
-   * We search the require cache against the known dependencies and
-   * devDependencies in the package definition (package.json), and look for an
-   * identical reference.
-   *
-   * Return the name of the module if found; false otherwise.
    */
-  getExternalModules (pkg) {
+  getExternalModules (pkg, rootPath) {
     if (!process.moduleLoadList) {
       throw new Error('process.moduleLoadList is not defined.')
     }
+
+    const modulePath = path.join(rootPath, 'node_modules')
     const dependencies = Object.keys(pkg.dependencies || { })
     const devDependencies = Object.keys(pkg.devDependencies || { })
     const nativeModules = process.moduleLoadList
@@ -95,10 +88,16 @@ module.exports = {
       .filter(id => !(/^NativeModule _/.test(id)))
       .filter(id => !(/\//g.test(id)))
       .map(id => /^NativeModule (\w+)$/.exec(id)[1])
-
-    return nativeModules
+    const requiredModules = Object.keys(require.cache).filter(mod => {
+      return mod.indexOf(modulePath) >= 0
+    })
+    const externalModules = nativeModules
+      .concat(requiredModules)
       .concat(dependencies)
       .concat(devDependencies)
+      .concat(requiredModules)
+
+    return Array.from(new Set(externalModules))
   },
 
 
@@ -111,7 +110,6 @@ module.exports = {
    * @param [pkg] the package definition to use for exceptions. optional.
    */
   freezeConfig (config, modules) {
-    modules || (modules = [ ])
     const propNames = Object.getOwnPropertyNames(config)
 
     propNames.forEach(name => {
@@ -122,7 +120,6 @@ module.exports = {
       }
 
       const ignoreModule = modules.find(moduleId => require(moduleId) === prop)
-
       if (ignoreModule) {
         return
       }
@@ -174,7 +171,9 @@ module.exports = {
    */
   bindApplicationEvents (app) {
     app.once('error', err => app.stop(err))
-    app.once('trailpack:all:configured', () => lib.Trails.freezeConfig(app.config, app.loadedModules))
+    app.once('trailpack:all:configured', () => {
+      lib.Trails.freezeConfig(app.config, app.loadedModules)
+    })
     app.once('trails:stop', () => lib.Trails.unfreezeConfig(app))
     app.once('trails:error:fatal', err => app.stop(err))
   },

--- a/lib/trails.js
+++ b/lib/trails.js
@@ -116,10 +116,17 @@ module.exports = {
 
     propNames.forEach(name => {
       const prop = config[name]
-      const ignoreModule = modules.find(moduleId => require(moduleId) === prop)
-      if (typeof prop === 'object' && prop !== null && !ignoreModule) {
-        lib.Trails.freezeConfig(prop, modules)
+
+      if (!prop || typeof prop !== 'object' || prop.constructor !== Object) {
+        return
       }
+
+      const ignoreModule = modules.find(moduleId => require(moduleId) === prop)
+
+      if (ignoreModule) {
+        return
+      }
+      lib.Trails.freezeConfig(prop, modules)
     })
     Object.freeze(config)
   },

--- a/lib/trails.js
+++ b/lib/trails.js
@@ -128,6 +128,7 @@ module.exports = {
       }
       lib.Trails.freezeConfig(prop, modules)
     })
+
     Object.freeze(config)
   },
 

--- a/test/index.js
+++ b/test/index.js
@@ -107,7 +107,7 @@ describe('Trails', () => {
           })
         })
         describe('@PackageNotDefinedError', () => {
-          it('should throw PackageNotDefinedError if no package.json definition is provided', () => {
+          it('should throw PackageNotDefinedError if no pkg definition is provided', () => {
             const def = {
               config: {
                 main: {
@@ -129,7 +129,7 @@ describe('Trails', () => {
             config: {
               main: { },
               log: {
-                logger: new smokesignals.Logger()
+                logger: new smokesignals.Logger('silent')
               }
             },
             pkg: { }
@@ -149,7 +149,7 @@ describe('Trails', () => {
               main: {
                 packs: [ smokesignals.Trailpack ]
               },
-              log: { logger: new smokesignals.Logger('debug') },
+              log: { logger: new smokesignals.Logger('silent') },
               foo: 'bar'
             }
           }
@@ -172,7 +172,7 @@ describe('Trails', () => {
               main: {
                 packs: [ smokesignals.Trailpack ]
               },
-              log: { logger: new smokesignals.Logger('debug') },
+              log: { logger: new smokesignals.Logger('silent') },
               foo: 'bar'
             }
           }

--- a/test/lib/trails.test.js
+++ b/test/lib/trails.test.js
@@ -199,11 +199,13 @@ describe('lib.Trails', () => {
     })
 
     // https://bugs.chromium.org/p/v8/issues/detail?id=4460
-    it('v8 issue 4460 exists', () => {
-      assert.throws(() => Object.freeze(new Int8Array()), TypeError)
-      //assert.throws(() => Object.freeze(new Buffer([1,2,3])), TypeError)
-      //assert.throws(() => Object.freeze(new DataView()), TypeError)
-    })
+    if (!/^v6/.test(process.version)) {
+      it('v8 issue 4460 exists', () => {
+        assert.throws(() => Object.freeze(new Int8Array()), TypeError)
+        //assert.throws(() => Object.freeze(new Buffer([1,2,3])), TypeError)
+        //assert.throws(() => Object.freeze(new DataView()), TypeError)
+      })
+    }
     it('should freeze objects containing unfreezable types without error', () => {
       const o1 = {
         typedArray: new Int8Array(),

--- a/test/lib/trails.test.js
+++ b/test/lib/trails.test.js
@@ -1,5 +1,6 @@
 'use strict'
 
+const path = require('path')
 const assert = require('assert')
 const smokesignals = require('smokesignals')
 const lib = require('../../lib')
@@ -128,7 +129,7 @@ describe('lib.Trails', () => {
       const testConfig = {
         main: { },
         log: {
-          logger: new smokesignals.Logger()
+          logger: new smokesignals.Logger('silent')
         },
         env: {
           envtest: {
@@ -151,7 +152,7 @@ describe('lib.Trails', () => {
       const testConfig = {
         main: { },
         log: {
-          logger: new smokesignals.Logger()
+          logger: new smokesignals.Logger('silent')
         },
         env: {
           envtest: {
@@ -166,7 +167,7 @@ describe('lib.Trails', () => {
       const testConfig = {
         main: { },
         log: {
-          logger: new smokesignals.Logger()
+          logger: new smokesignals.Logger('silent')
         },
         env: {
           env: 'hello'
@@ -180,7 +181,7 @@ describe('lib.Trails', () => {
   describe('#freezeConfig', () => {
     it('should freeze nested object', () => {
       const o1 = { foo: { bar: 1 } }
-      lib.Trails.freezeConfig(o1)
+      lib.Trails.freezeConfig(o1, [ ])
 
       assert.throws(() => o1.foo = null, Error)
     })
@@ -225,12 +226,35 @@ describe('lib.Trails', () => {
           foo: 'bar'
         }
       }
-      lib.Trails.freezeConfig(app.config)
+      lib.Trails.freezeConfig(app.config, [ ])
       assert.throws(() => app.config.a = 2, Error)
 
       lib.Trails.unfreezeConfig(app)
       app.config.a = 2
       assert.equal(app.config.a, 2)
+    })
+  })
+
+  describe('#getExternalModules', () => {
+    const rootPath = path.resolve(path.dirname(require.main.id))
+    const pkg = {
+      dependencies: {
+      },
+      devDependencies: {
+        smokesignals: '0.0.0',
+        mocha: '0.0.0'
+      }
+    }
+
+    it('should return native modules', () => {
+      const modules = lib.Trails.getExternalModules(pkg, rootPath)
+      assert(modules.indexOf('path') !== -1)
+      assert(modules.indexOf('events') !== -1)
+    })
+    it('should return external modules', () => {
+      const modules = lib.Trails.getExternalModules(pkg, rootPath)
+      assert(modules.indexOf('smokesignals') !== -1)
+      assert(modules.indexOf('mocha') !== -1)
     })
   })
 

--- a/test/lib/trails.test.js
+++ b/test/lib/trails.test.js
@@ -196,6 +196,25 @@ describe('lib.Trails', () => {
       o1.foo.x = 1
       assert.equal(o1.foo.x, 1)
     })
+
+    // https://bugs.chromium.org/p/v8/issues/detail?id=4460
+    it('v8 issue 4460 exists', () => {
+      assert.throws(() => Object.freeze(new Int8Array()), TypeError)
+      //assert.throws(() => Object.freeze(new Buffer([1,2,3])), TypeError)
+      //assert.throws(() => Object.freeze(new DataView()), TypeError)
+    })
+    it('should freeze objects containing unfreezable types without error', () => {
+      const o1 = {
+        typedArray: new Int8Array(),
+        buffer: new Buffer([ 1,2,3 ]),
+        fun: function () { }
+      }
+      lib.Trails.freezeConfig(o1, [ ])
+
+      assert(o1.typedArray)
+      assert(Buffer.isBuffer(o1.buffer))
+      assert(o1.fun)
+    })
   })
 
   describe('#unfreezeConfig', () => {

--- a/test/testapp.js
+++ b/test/testapp.js
@@ -1,3 +1,5 @@
+const smokesignals = require('smokesignals')
+
 module.exports = {
   api: {
 
@@ -8,7 +10,9 @@ module.exports = {
         root: __dirname
       }
     },
-    log: require('../archetype/config/log')
+    log: {
+      logger: new smokesignals.Logger('silent')
+    }
   },
   pkg: {
 

